Crash in AuxiliaryProcessProxy::connection under WebPageProxy::sendWheelEvent

mcatanzaro opened the pull request:
#### 7c79d4d8b355b0f8813ab418f885046ffb2defba
<pre>
Crash in AuxiliaryProcessProxy::connection under WebPageProxy::sendWheelEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=295679">https://bugs.webkit.org/show_bug.cgi?id=295679</a>

Reviewed by Chris Dumez.

This is a speculative fix for the crash. The user may generate a wheel
event at any time, including after the web process is gone. All of our
other event handlers check for a running web process before sending
these messages, but the checks in WebPageProxy::sendWheelEvent seem
inadequate.

Canonical link: <a href="https://commits.webkit.org/302030@main">https://commits.webkit.org/302030@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebd6312716761644d94be9100b1ba6460e6a7ac3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127447 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134523 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79002 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47711 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55619 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97018 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64969 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130395 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38171 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114147 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77498 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37004 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77897 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108030 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32672 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137008 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54106 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41693 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105540 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54617 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110496 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105193 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26897 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50704 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29162 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51685 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54043 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60130 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53277 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56734 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55036 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4243,6 +4243,9 @@ void WebPageProxy::sendWheelEvent(WebCore::FrameIdentifier frameID, const WebWhe
     internals().wheelEventActivityHysteresis.impulse();
 #endif
 
+    if (!hasRunningProcess())
```
Can we write an API test for this?

cdumez replied:
I could understand returning early if the page is closed. However, what's the reasoning behind returning early when there is no running process? Shouldn't we call `ensureRunningProcess()` and launch one if the page is not closed and there is no running process (page is still using dummy process or the process was terminated).

mcatanzaro replied:
> Can we write an API test for this?

Hm, I'm not sure:

 * It might be possible to do a GTK API level test using `gdk_display_put_event()`, but this is deprecated and will be removed in GTK 5, so doesn't seem worth spending time on. I can't think of another way to inject an event without JavaScript, which we cannot use because there needs to be no running process. Also, we'd need a test function for causing the process to crash, which doesn't seem very nice. This approach doesn't seem worth it.
 * But maybe I can find a way to do a cross-platform test, since there we have more control and should be able to just inject events directly. I'll see if it can be done without too much effort. I don't plan to spend a lot of time on this.

> I could understand returning early if the page is closed. However, what's the reasoning behind returning early when there is no running process? Shouldn't we call ensureRunningProcess() and launch one if the page is not closed and there is no running process (page is still using dummy process or the process was terminated).

What I actually want to do is return early if there is no IPC::Connection, so I originally wrote `if (!hasConnection())`. But it so happens that is only possible if there is no running process, so it's OK to check for running process instead. WebPageProxy already does so in about a hundred other places, including all the other event handling functions (`WebPageProxy::handleKeyboardEvent`, `WebPageProxy::handleMouseEvent`, `WebPageProxy::handleTouchEvent`, and indeed even `WebPageProxy::handleWheelEvent`), whereas it checks for an active connection in relatively few places, so I selected this way in order to match the existing code.

mcatanzaro replied:
> But maybe I can find a way to do a cross-platform test, since there we have more control and should be able to just inject events directly.

Our cross-platform tests use the C API, which doesn't provide a way to synthesize a wheel event. I would need access to WebPageProxy directly. I know it's important to add tests when we can, but I think it's not worth the effort this time.

cdumez reviewed (approved):
Ok.

mcatanzaro committed:
Crash in AuxiliaryProcessProxy::connection under WebPageProxy::sendWheelEvent
https://bugs.webkit.org/show_bug.cgi?id=295679

Reviewed by Chris Dumez.

This is a speculative fix for the crash. The user may generate a wheel
event at any time, including after the web process is gone. All of our
other event handlers check for a running web process before sending
these messages, but the checks in WebPageProxy::sendWheelEvent seem
inadequate.

Canonical link: https://commits.webkit.org/302030@main
